Add parameter validation for ARIA modules

hanno-arm opened the pull request:
__Summary:__ Based on #1643, this PR adds parameter validation for the ARIA module.

__Work left:__ Allocate free error code for `MBEDTLS_ERR_ARIA_BAD_INPUT_DATA`. Slightly annoying because there are already two similar error codes `MBEDTLS_ERR_ARIA_INVALID_KEY_LENGTH` and `MBEDTLS_ERR_ARIA_INVALID_INPUT_LENGTH`, and we're running out of low level error codes.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -85,29 +88,32 @@ mbedtls_aria_context;
  *                 the context.
  *
- * \param ctx      The ARIA context to initialize.
+ * \param ctx      The ARIA context to initialize. Must not be \c NULL.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -49,6 +49,9 @@
 #define MBEDTLS_ERR_ARIA_INVALID_KEY_LENGTH   -0x005C  /**< Invalid key length. */
 #define MBEDTLS_ERR_ARIA_INVALID_INPUT_LENGTH -0x005E  /**< Invalid data input length. */
 
+/** TEMPORARY -- THIS IS IN CONFLICT WITH EXISTING ERROR CODES AND NEEDS CHANGE. */
```
Temp remaining

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -85,29 +88,32 @@ mbedtls_aria_context;
  * \brief          This function releases and clears the specified ARIA context.
  *
- * \param ctx      The ARIA context to clear.
+ * \param ctx      The ARIA context to clear. May be \c NULL, in which
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -85,29 +88,32 @@ mbedtls_aria_context;
  *
  * \param ctx      The ARIA context to which the key should be bound.
- * \param key      The encryption key.
+ *                 Must be initialized.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -85,29 +88,32 @@ mbedtls_aria_context;
  * \param ctx      The ARIA context to which the key should be bound.
- * \param key      The encryption key.
+ *                 Must be initialized.
+ * \param key      The encryption key. Must be a readable buffer
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -85,29 +88,32 @@ mbedtls_aria_context;
- * \param key      The encryption key.
+ *                 Must be initialized.
+ * \param key      The encryption key. Must be a readable buffer
+ *                 of size \p keybits bits.
```
Bits?

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -117,13 +123,16 @@ int mbedtls_aria_setkey_enc( mbedtls_aria_context *ctx,
  *
  * \param ctx      The ARIA context to which the key should be bound.
- * \param key      The decryption key.
+ *                 Must be initialized.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -117,13 +123,16 @@ int mbedtls_aria_setkey_enc( mbedtls_aria_context *ctx,
  * \param ctx      The ARIA context to which the key should be bound.
- * \param key      The decryption key.
+ *                 Must be initialized.
+ * \param key      The decryption key. Must be a readable buffer
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -142,10 +151,12 @@ int mbedtls_aria_setkey_dec( mbedtls_aria_context *ctx,
  *                 call to this API with the same context.
  *
  * \param ctx      The ARIA context to use for encryption or decryption.
+ *                 Must be initialized.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -177,16 +188,20 @@ int mbedtls_aria_crypt_ecb( mbedtls_aria_context *ctx,
  *
  *
  * \param ctx      The ARIA context to use for encryption or decryption.
+ *                 Must be initialized.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -177,16 +188,20 @@ int mbedtls_aria_crypt_ecb( mbedtls_aria_context *ctx,
  * \param length   The length of the input data in Bytes. This must be a
  *                 multiple of the block size (16 Bytes).
  * \param iv       Initialization vector (updated after use).
+ *                 Must be a readable buffer of size 16 Bytes.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -177,16 +188,20 @@ int mbedtls_aria_crypt_ecb( mbedtls_aria_context *ctx,
  * \param iv       Initialization vector (updated after use).
+ *                 Must be a readable buffer of size 16 Bytes.
  * \param input    The buffer holding the input data.
+ *                 May be \c NULL if `length == 0`.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -177,16 +188,20 @@ int mbedtls_aria_crypt_ecb( mbedtls_aria_context *ctx,
  * \param input    The buffer holding the input data.
+ *                 May be \c NULL if `length == 0`.
  * \param output   The buffer holding the output data.
+ *                 May be \c NULL if `length == 0`.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -221,15 +236,21 @@ int mbedtls_aria_crypt_cbc( mbedtls_aria_context *ctx,
  *
  *
  * \param ctx      The ARIA context to use for encryption or decryption.
+ *                 Must be initialized.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -221,15 +236,21 @@ int mbedtls_aria_crypt_cbc( mbedtls_aria_context *ctx,
  *                 #MBEDTLS_ARIA_DECRYPT.
  * \param length   The length of the input data.
  * \param iv_off   The offset in IV (updated after use).
+ *                 Must not be larger than 15.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -221,15 +236,21 @@ int mbedtls_aria_crypt_cbc( mbedtls_aria_context *ctx,
  * \param iv_off   The offset in IV (updated after use).
+ *                 Must not be larger than 15.
  * \param iv       The initialization vector (updated after use).
+ *                 Must be a readable buffer of size 16 Bytes.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -221,15 +236,21 @@ int mbedtls_aria_crypt_cbc( mbedtls_aria_context *ctx,
  * \param iv       The initialization vector (updated after use).
+ *                 Must be a readable buffer of size 16 Bytes.
  * \param input    The buffer holding the input data.
+ *                 May be \c NULL if `length == 0`.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -221,15 +236,21 @@ int mbedtls_aria_crypt_cbc( mbedtls_aria_context *ctx,
  * \param input    The buffer holding the input data.
+ *                 May be \c NULL if `length == 0`.
  * \param output   The buffer holding the output data.
+ *                 May be \c NULL if `length == 0`.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -299,17 +320,24 @@ int mbedtls_aria_crypt_cfb128( mbedtls_aria_context *ctx,
  *             securely discarded as soon as it's no longer needed.
  *
  * \param ctx              The ARIA context to use for encryption or decryption.
+ *                         Must be initialized.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -299,17 +320,24 @@ int mbedtls_aria_crypt_cfb128( mbedtls_aria_context *ctx,
- * \param nonce_counter    The 128-bit nonce and counter.
- * \param stream_block     The saved stream block for resuming. This is
- *                         overwritten by the function.
+ *                         Must not be larger than 15.
```
This*

hanno-arm replied:
Fixed.

AndrzejKurek commented on include/mbedtls/aria.h:
```diff
@@ -299,17 +320,24 @@ int mbedtls_aria_crypt_cfb128( mbedtls_aria_context *ctx,
+ *                         an RW-buffer of length 16 bytes.
+ *                         This is overwritten by the function.
  * \param input            The buffer holding the input data.
+ *                         May be \c NULL if `length == 0`.
```
It? (here and below)

hanno-arm replied:
Fixed.

hanno-arm commented:
Thanks @AndrzejKurek for your comments. I have fixed the issues you found and some further minor improvements. 

The error code handling is still pending, but could you already check if you're ok with the rest? Thanks!

sbutcher-arm commented on include/mbedtls/aria.h:
```diff
@@ -85,29 +88,33 @@ mbedtls_aria_context;
  *
- * \param ctx      The ARIA context to clear.
+ * \param ctx      The ARIA context to clear. This may be \c NULL, in which
+ *                 case this function is a no-op. If it is not \c NULL,
```
'no-op' isn't a real word.

hanno-arm replied:
Ok.

sbutcher-arm commented on include/mbedtls/aria.h:
```diff
@@ -299,17 +323,24 @@ int mbedtls_aria_crypt_cfb128( mbedtls_aria_context *ctx,
- * \param nonce_counter    The 128-bit nonce and counter.
- * \param stream_block     The saved stream block for resuming. This is
- *                         overwritten by the function.
+ *                         This must not be larger than 15.
```
We shouldn't just give a number, but number and unit. Therefore, shouldn't it be:
> This must not be larger than 15 bytes.

hanno-arm replied:
In the end, you pass a unit-less number, so I would have preferred not to add this here, but I'm also ok changing it.

sbutcher-arm commented on include/mbedtls/aria.h:
```diff
@@ -299,17 +323,24 @@ int mbedtls_aria_crypt_cfb128( mbedtls_aria_context *ctx,
- *                         overwritten by the function.
+ *                         This must not be larger than 15.
+ * \param nonce_counter    The 128-bit nonce and counter. This must point to
+ *                         an RW-buffer of length 16 bytes.
```
I don't think we're allowed to say 'RW-buffer'. This should be 'read/write buffer'.

hanno-arm replied:
Ok.

sbutcher-arm commented on include/mbedtls/aria.h:
```diff
@@ -299,17 +323,24 @@ int mbedtls_aria_crypt_cfb128( mbedtls_aria_context *ctx,
+ * \param nonce_counter    The 128-bit nonce and counter. This must point to
+ *                         an RW-buffer of length 16 bytes.
+ * \param stream_block     The saved stream block for resuming. This must
+ *                         point to an RW-buffer of length 16 bytes.
```
'RW-buffer' again.

sbutcher-arm commented on tests/suites/test_suite_aria.function:
```diff
@@ -16,6 +16,185 @@
+                            mbedtls_aria_setkey_enc( NULL, key,
+                                                     sizeof( key ) ) );
+    TEST_INVALID_PARAM_RET( MBEDTLS_ERR_ARIA_BAD_INPUT_DATA,
+                           mbedtls_aria_setkey_enc( &ctx, NULL,
```
Indentation.

hanno-arm replied:
Fixing.

sbutcher-arm commented on tests/suites/test_suite_aria.function:
```diff
@@ -16,6 +16,185 @@
+                                                    sizeof( key ) ) );
+
+    TEST_INVALID_PARAM_RET( MBEDTLS_ERR_ARIA_BAD_INPUT_DATA,
+                           mbedtls_aria_setkey_dec( NULL, key,
```
Indentation.

hanno-arm replied:
Fixing.

sbutcher-arm commented on tests/suites/test_suite_aria.function:
```diff
@@ -16,6 +16,185 @@
+                           mbedtls_aria_setkey_dec( NULL, key,
+                                                    sizeof( key ) ) );
+    TEST_INVALID_PARAM_RET( MBEDTLS_ERR_ARIA_BAD_INPUT_DATA,
+                           mbedtls_aria_setkey_dec( &ctx, NULL,
```
Indentation.

hanno-arm replied:
Fixing.

hanno-arm commented:
@sbutcher-arm Thanks for your review, I think I have addressed your comments. Specifically, apart from wording and style changes, I added a test for `mbedtls_aes_free()` to the new parameter validation test. I didn't change the selftest because I think that's out of scope of parameter validation.

sbutcher-arm commented on tests/suites/test_suite_aria.function:
```diff
@@ -16,6 +16,186 @@
+    size_t iv_off = 0;
+
+    TEST_INVALID_PARAM( mbedtls_aria_init( NULL ) );
+    TEST_VALID_PARAM( mbedtls_aria_free( NULL ) );
```
The test of `TEST_VALID_PARAM( mbedtls_aria_free( NULL ) )` has it's own function in AES because it doesn't depend on `MBEDTLS_CHECK_PARAMS` and `!MBEDTLS_PARAM_FAILED_ALT`.

If that logic applies there, it should also apply here.

hanno-arm replied:
It does. I deliberately kept it in the `MBEDTLS_CHECK_PARAMS`-only test `invalid_params()`, but if you prefer I can add a separate test.

hanno-arm replied:
I will add a separate test in line with the AES PR.

sbutcher-arm commented:
Jenkins CI is failing on the BSD timing test. Good enough for a pass.

sbutcher-arm commented:
Jenkins CI is failing only on FreeBSD timing test which is as good as pass. Therefore ready to merge!
